pkg/channel/event_dispatcher.go

```diff
@@ -0,0 +1,165 @@
+}
+
+// NewEventDispatcher creates a new event dispatcher that can dispatch
+// events to HTTP destinations.
+func NewEventDispatcher(logger *zap.Logger) *EventDispatcher {
+	ceClient, err := kncloudevents.NewDefaultClient()
+	if err != nil {
+		logger.Fatal("failed to create cloudevents client", zap.Error(err))
+	}
+	return &EventDispatcher{
+		ceClient:         ceClient,
+		supportedSchemes: sets.NewString("http", "https"),
+		logger:           logger,
+	}
+}
+
+// DispatchEvent dispatches an event to a destination over HTTP.
+//
+// The destination and reply are DNS names. For names with a single label,
+// the default namespace is used to expand it into a fully qualified name
+// within the cluster.
+func (d *EventDispatcher) DispatchEvent(ctx context.Context, event cloudevents.Event, destination, reply string, defaults DispatchDefaults) error {
+	var err error
+	// Default to replying with the original event. If there is a destination, then replace it
+	// with the response from the call to the destination instead.
+	response := &event
+	if destination != "" {
+		destinationURL := d.resolveURL(destination, defaults.Namespace)
+		ctx, response, err = d.executeRequest(ctx, destinationURL, event)
+		if err != nil {
+			return fmt.Errorf("unable to complete request to %s: %v", destinationURL, err)
+		}
+	}
+
+	if reply != "" && response != nil {
+		replyURL := d.resolveURL(reply, defaults.Namespace)
+		_, _, err = d.executeRequest(ctx, replyURL, *response)
+		if err != nil {
+			return fmt.Errorf("failed to forward reply to %s: %v", replyURL, err)
+		}
+	}
+	return nil
+}
+
+func (d *EventDispatcher) executeRequest(ctx context.Context, url *url.URL, event cloudevents.Event) (context.Context, *cloudevents.Event, error) {
+	d.logger.Debug("Dispatching event", zap.String("url", url.String()))
+
+	tctx := cloudevents.HTTPTransportContextFrom(ctx)
+	sctx := utils.ContextFrom(tctx, url)
+	sctx = addOutGoingTracing(sctx, url)
+
+	rctx, reply, err := d.ceClient.Send(sctx, event)
+	if err != nil {
+		return rctx, nil, err
+	}
+	rtctx := cloudevents.HTTPTransportContextFrom(rctx)
+	if isFailure(rtctx.StatusCode) {
+		// reject non-successful responses
+		return rctx, nil, fmt.Errorf("unexpected HTTP response, expected 2xx, got %d", rtctx.StatusCode)
+	}
+	headers := utils.PassThroughHeaders(rtctx.Header)
+	if correlationID, ok := tctx.Header[correlationIDHeaderName]; ok {
+		headers[correlationIDHeaderName] = correlationID
+	}
+	rtctx.Header = http.Header(headers)
+	rctx = cehttp.WithTransportContext(rctx, rtctx)
+	return rctx, reply, nil
+}
+
+func addOutGoingTracing(ctx context.Context, url *url.URL) context.Context {
+	tctx := cloudevents.HTTPTransportContextFrom(ctx)
+	// Creating a dummy request to leverage propagation.SpanContextFromRequest method
+	req := &http.Request{
+		Header: tctx.Header,
+	}
+	// Attach the Span context that is currently saved in the request's headers.
+	if sc, ok := propagation.SpanContextFromRequest(req); ok {
```

Comment:
I think that this creates an extra, unreported span. #1944 improved on this a bit, but assumes that there is a *Span resident in memory, which works for InMemory, but probably not other channel implementations. 

We might need to use the [traceparent](https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md#traceparent) event attribute for serialization into the CloudEvent itself.

Reply:
OK, yes we will need to figure this one out. Let's chat in person more

Reply:
Agreed we will do a follow up PR